cavern/src/main/java/org/opencadc/cavern/files/CavernURLGenerator.java

```diff
@@ -211,56 +231,45 @@ private List<URI> handleDataNode(VOSURI target, DataNode node, Protocol protocol
                 case VOS.PROTOCOL_HTTPS_GET:
                     scheme = "https";
                     dir = Direction.pullFromVoSpace;
+                    grantClass = ReadGrant.class;
                     break;
                 case VOS.PROTOCOL_HTTPS_PUT:
                     scheme = "https";
                     dir = Direction.pushToVoSpace;
+                    grantClass = WriteGrant.class;
                     break;
-            }
+           }
 
             List<URL> baseURLs = getBaseURLs(target, protocol.getSecurityMethod(), scheme);
             if (baseURLs == null || baseURLs.isEmpty()) {
                 log.debug("no matching interfaces ");
                 return new ArrayList<URI>(0);
             }
-            
+
             if (dir == null) {
                 log.debug("no matching protocols");
                 return new ArrayList<URI>(0);
             }
 
-            // create the metadata and signature segments
-            StringBuilder metaSb = new StringBuilder();
-            metaSb.append(KEY_META_NODE).append("=").append(target.toString());
-            metaSb.append("&");
-            metaSb.append(KEY_META_DIRECTION).append("=").append(dir.getValue());
-            byte[] metaBytes = metaSb.toString().getBytes();
-
-            RsaSignatureGenerator sg = new RsaSignatureGenerator(PRIV_KEY_FILENAME);
-            String sig;
-            try {
-                byte[] sigBytes = sg.sign(new ByteArrayInputStream(metaBytes));
-                sig = new String(Base64.encode(sigBytes));
-                log.debug("Created signature: " + sig + " for meta: " + metaSb.toString());
-            } catch (InvalidKeyException | IOException | RuntimeException e) {
-                throw new IllegalStateException("Could not sign url", e);
-            }
-            String meta = new String(Base64.encode(metaBytes));
-            log.debug("meta: " + meta);
-            log.debug("sig: " + sig);
+            // Use TokenTool to generate a preauth token
+            File privateKeyFile = findFile(PRIV_KEY_FILENAME);
+            File pubKeyFile = findFile(PUB_KEY_FILENAME);
 
-            // build the request path
-            StringBuilder path = new StringBuilder();
-            String metaURLEncoded = base64URLEncode(meta);
-            String sigURLEncoded = base64URLEncode(sig);
+            TokenTool gen = new TokenTool(pubKeyFile, privateKeyFile);
 
-            log.debug("metaURLEncoded: " + metaURLEncoded);
-            log.debug("sigURLEncoded: " + sigURLEncoded);
+            // Format of token is <base64 url encoded meta>.<base64 url encoded signature>
+            PosixIdentityManager im = new PosixIdentityManager(FileSystems.getDefault().getUserPrincipalLookupService());
+            String callingUser = (String)im.toOwner(s);
```

Comment:
What happens here if the call is made anonymously?  Is callingUser null?
There's a method in AuthenticationUtil called `getUseridsFromSubject` which might work better for this.  You'd take the first Principal in the Set and use the getName() method to get the userid.

Reply:
done